listmatch/listmatch.py

```diff
@@ -42,19 +41,27 @@ def __init__(self, start, end):
         end.is_end = True
 
     def match(self, target):
-        states = StateSet([self.start])
+        states = StateSet([(self.start, [[]])])
 
         for elem in target:
-            states = StateSet(
-                state.next_state
-                for state in states
-                if state.matcher(elem)
-            )
-
-        for s in states:
+            next_state_list = []
+            for state, paths in states:
+                ret = state.matcher(elem)
+                matched, aux = ret if isinstance(ret, tuple) else (ret, None)
```

Comment:
small naming rant @pokey : if `aux` and `path` are same thing/compatible (e.g. debug info) might want to try and name them more similarly?